lib/Collaboration/Reference/TalkReferenceProvider.php

```diff
@@ -0,0 +1,288 @@
+use OCP\IL10N;
+use OCP\IURLGenerator;
+
+/**
+ * @psalm-type ReferenceMatch = array{token: string, message: ?int}
+ */
+class TalkReferenceProvider implements IReferenceProvider {
+	protected IURLGenerator $urlGenerator;
+	protected Manager $roomManager;
+	protected ChatManager $chatManager;
+	protected MessageParser $messageParser;
+	protected IL10N $l;
+	protected ?string $userId;
+
+	public function __construct(IURLGenerator $urlGenerator,
+								Manager $manager,
+								ChatManager $chatManager,
+								MessageParser $messageParser,
+								IL10N $l,
+								?string $userId) {
+		$this->urlGenerator = $urlGenerator;
+		$this->roomManager = $manager;
+		$this->chatManager = $chatManager;
+		$this->messageParser = $messageParser;
+		$this->l = $l;
+		$this->userId = $userId;
+	}
+
+
+	public function matchReference(string $referenceText): bool {
+		return $this->getTalkAppLinkToken($referenceText) !== null;
+	}
+
+	/**
+	 * @param string $referenceText
+	 * @return array|null
+	 * @psalm-return ReferenceMatch|null
+	 */
+	private function getTalkAppLinkToken(string $referenceText): ?array {
+		$indexPhpUrl = $this->urlGenerator->getAbsoluteURL('/index.php/call/');
+		$rewriteUrl = $this->urlGenerator->getAbsoluteURL('/call/');
+
+		if (str_starts_with($referenceText, $indexPhpUrl)) {
+			$urlOfInterest = substr($referenceText, strlen($indexPhpUrl)) ?: null;
+		} elseif (str_starts_with($referenceText, $rewriteUrl)) {
+			$urlOfInterest = substr($referenceText, strlen($rewriteUrl)) ?: null;
+		} else {
+			return null;
+		}
+
+		$hashPosition = strpos($urlOfInterest, '#');
+		$queryPosition = strpos($urlOfInterest, '?');
+
+		if ($hashPosition === false && $queryPosition === false) {
+			return [
+				'token' => $urlOfInterest,
+				'message' => null,
+			];
+		}
+
+		if ($hashPosition !== false && $queryPosition !== false) {
+			$cutPosition = min($hashPosition, $queryPosition);
+		} elseif ($hashPosition !== false) {
+			$cutPosition = $hashPosition;
+		} else {
+			$cutPosition = $queryPosition;
+		}
+
+		$token = substr($urlOfInterest, 0, $cutPosition);
+		$messageId = null;
+		if ($hashPosition !== false) {
+			$afterHash = substr($urlOfInterest, $hashPosition + 1);
+			if (preg_match('/^message_(\d+)$/', $afterHash, $matches)) {
+				$messageId = $matches[1];
+			}
+		}
+
+		return [
+			'token' => $token,
+			'message' => $messageId,
+		];
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	public function resolveReference(string $referenceText): ?IReference {
+		if ($this->matchReference($referenceText)) {
+			$reference = new Reference($referenceText);
+			try {
+				$this->fetchReference($reference);
+			} catch (RoomNotFoundException|ParticipantNotFoundException $e) {
+				$reference->setRichObject('call', null);
+				$reference->setAccessible(false);
+			}
+			return $reference;
+		}
+
+		return null;
+	}
+
+	/**
+	 * @throws RoomNotFoundException
+	 */
+	private function fetchReference(Reference $reference): void {
```

Comment:
What you think to use a tool to check the cyclomatic complexity in Talk? Like this, we have a lot of other big methods.